maitake/src/task/storage.rs

```diff
@@ -0,0 +1,74 @@
+use super::Task;
+use core::future::Future;
+use core::ptr::NonNull;
+
+/// A trait representing the storage medium of a [`Task`].
+///
+/// This is used to contain tasks at runtime, and abstract over the
+/// type erasure and type recovery steps, e.g. converting a heap
+/// allocation type into a NonNull pointer, and recovering it
+/// back into a heap allocation from a NonNull pointer
+///
+/// This trait is exposed publicly to allow for end users to implement
+/// it in their own heap allocation types, if they are not specifically
+/// a [`Box`].
+///
+/// This trait is ONLY appropriate for heap allocation types that represent
+/// exclusive ownership of the contained data, as it may be mutated while
+/// in pointer form. For example, the `Box` type would match this guarantee,
+/// but the `Arc` type would not, as it allows for shared access, and NOT
+/// exclusive mutable access.
```

Comment:
(also, i'm not actually sure if this is necessarily a required invariant, since the task type is interior-ly mutable; the future lives in an `UnsafeCell` and only polling or dropping the task will mutate the `UnsafeCell`; the atomic state variable ensures mutual exclusion of the mutable part of the task. an earlier design used `Arc`s rather than `Box` but i abandoned that approach because i wanted to roll the ref count into the same atomic as the task's state flags...)